cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -891,17 +791,20 @@ private void validateAndCorrectRunningRunRecords() {
   void validateAndCorrectRunningRunRecords(final ProgramType programType,
                                            final Set<String> processedInvalidRunRecordIds) {
     final Map<RunId, RuntimeInfo> runIdToRuntimeInfo = runtimeService.list(programType);
+    com.google.common.base.Predicate<RunRecordMeta> notRunningPredicate =
+      new com.google.common.base.Predicate<RunRecordMeta>() {
+        @Override
+        public boolean apply(RunRecordMeta input) {
+          String runId = input.getPid();
+          // Check if it is not actually running.
+          return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
+        }
+      };
 
     LOG.trace("Start getting run records not actually running ...");
-    Collection<RunRecordMeta> notActuallyRunning = store.getRuns(ProgramRunStatus.RUNNING,
-                                                           new com.google.common.base.Predicate<RunRecordMeta>() {
-      @Override
-      public boolean apply(RunRecordMeta input) {
-        String runId = input.getPid();
-        // Check if it is not actually running.
-        return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
-      }
-    }).values();
+    Collection<RunRecordMeta> notActuallyRunning = store.getRuns(ProgramRunStatus.RUNNING, notRunningPredicate)
+                                                        .values();
+    notActuallyRunning.addAll(store.getRuns(ProgramRunStatus.STARTING, notRunningPredicate).values());
```

Comment:
Updated the RunRecordCorrector to look for both RUNNING or STARTING programs that are not actually running to set them to an ERROR status.

This works because the runtimeInfo will be added now just after the runtimeinfo is created, rather than [in the controller's listener after it has already been marked as ALIVE](https://github.com/caskdata/cdap/blob/develop/cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java#L396)